code/controllers/subsystem/processing/overlays.dm

```diff
@@ -0,0 +1,115 @@
+	currentrun = null
+
+/datum/subsystem/processing/overlays/New()
+	NEW_SS_GLOBAL(SSoverlays)
+
+/datum/subsystem/processing/overlays/Recover()
+	processing = SSoverlays.processing
+
+/datum/subsystem/processing/overlays/fire()
+	while(processing.len)
+		var/atom/thing = processing[processing.len]
+		processing.len--
+		if(thing)
+			if(LAZYLEN(thing.priority_overlays) && LAZYLEN(thing.our_overlays))
+				thing.overlays = (thing.our_overlays | thing.priority_overlays)
+			else if(LAZYLEN(thing.our_overlays))
+				thing.overlays = thing.our_overlays.Copy()
+			else if(LAZYLEN(thing.priority_overlays))			//Do these two have to be copied? We don't want assignments to them to trigger overlay updates. But, idk, bruh, BYOND. - Cyberboss
+				thing.overlays = thing.priority_overlays.Copy()
+			else
+				LAZYCLEARLIST(thing.overlays)
+		if(MC_TICK_CHECK)
+			break
+	can_fire = processing.len > 0
+
+/atom/proc/cut_overlays(priority = FALSE)
+	var/list/cached_overlays = our_overlays
+	var/list/cached_priority = priority_overlays
+	
+	var/need_compile = FALSE
+
+	if(LAZYLEN(cached_overlays)) //don't queue empty lists, don't cut priority overlays
+		cached_overlays.Cut()  //clear regular overlays
+		need_compile = TRUE
+
+	if(priority && LAZYLEN(cached_priority))
+		cached_priority.Cut()
+		need_compile = TRUE
+
+	if(need_compile)
+		SSoverlays.processing[src] = src
+		SSoverlays.can_fire = TRUE
+
+/atom/proc/cut_overlay(image, priority)
+	if(!image)
+		return
+	var/list/cached_overlays = our_overlays	//sanic
+	var/list/cached_priority = priority_overlays
+	var/init_o_len = LAZYLEN(cached_overlays)
+	var/init_p_len = LAZYLEN(cached_priority)  //starter pokemon
+	
+	LAZYREMOVE(cached_overlays, image)
+	if(priority)
+		LAZYREMOVE(cached_priority, image)
+
+	if((init_o_len != LAZYLEN(cached_priority)) || (init_p_len != LAZYLEN(cached_overlays)))
+		SSoverlays.processing[src] = src	//something was removed, queue it for compiling
+		SSoverlays.can_fire = TRUE
+
+/atom/proc/add_overlay(image, priority = FALSE)
+	if(!image)
+		return
+	LAZYINITLIST(our_overlays)	//always initialized after this point
+	LAZYINITLIST(priority_overlays)
+	var/list/cached_overlays = our_overlays	//sanic
+	var/list/cached_priority = priority_overlays
+	var/init_o_len = cached_overlays.len
+	var/init_p_len = cached_priority.len  //starter pokemon
+	var/need_compile
+
+	if(priority)
+		cached_priority |= image  //or in the image. Can we use [image] = image?
+		need_compile = init_p_len != cached_priority.len
+	else
+		cached_overlays |= image
+		need_compile = init_o_len != cached_overlays.len
+
+	if(need_compile) //have we caught more pokemon?
+		SSoverlays.processing[src] = src	//something was added, queue it for compiling
+		SSoverlays.can_fire = TRUE
+
+/atom/proc/copy_overlays(atom/other, cut_everything = FALSE)
+	if(!other)
+		return
+	copy_overlays_list(other.our_overlays, FALSE, FALSE)	//lets face it, we probably dont WANT priority overlays here (fire, x4, acid, etc..)
```

Comment:
could you add some more comments explaining what this procs are for, i'm having trouble visualizing it.